src/contact.rs

```diff
@@ -36,6 +37,51 @@ use crate::{chat, stock_str};
 /// Time during which a contact is considered as seen recently.
 const SEEN_RECENTLY_SECONDS: i64 = 600;
 
+/// Valid contact address.
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct ContactAddress<'a>(&'a str);
+
+impl Deref for ContactAddress<'_> {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+
+impl AsRef<str> for ContactAddress<'_> {
+    fn as_ref(&self) -> &str {
+        self.0
+    }
+}
+
+impl fmt::Display for ContactAddress<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl<'a> ContactAddress<'a> {
+    /// Constructs a new contact address from string,
+    /// normalizing and validating it.
+    pub fn new(s: &'a str) -> Result<Self> {
+        let addr = addr_normalize(s);
+        if !may_be_valid_addr(addr) {
+            bail!("invalid address {:?}", s);
```

Comment:
Btw, can't we have Context as a per-thread global (TLS) var so that errors could be logged where they are created? I saw such an approach in some C projects